### DISCUSSION TITLE
Example BMI extension implementations

### DISCUSSION THREAD

**Discussion by mdpiper** [csdms/bmi#140] [Ideas]
To help make the notion of a [BMI extension](https://github.com/csdms/bmi/discussions/138) more tangible, @mcflugen created a sample implementation of a geospatial BMI extension in Python (https://github.com/csdms/bmi-example-python/pull/31) and I followed him with one in Fortran (https://github.com/csdms/bmi-example-fortran/pull/20).

**Comment by mdpiper** [comment]
Please let us know what you think about these examples, either here or in the respective PRs.

**Comment by hellkite500** [comment]
I finally got some time to shore up some thoughts on this and build some simple functionality, examples, and tests.  I have created these in a [C/C++ BMI extension library](https://github.com/hellkite500/bmi-c-extensions) here to show some possible methods for how to technically implement extensions.  This doesn't have any of the SIDL components, but is instead intended to show how to get similar functionality and even patterns similar to existing BMI with unique extensions which don't change the core BMI in any way.  There is an example BMI model in the repo, and unit test codes testing several possible solutions to managing/implementing extensions in this way.

I also have some python prototypes that I would say are "similar" but leverage some different OO strategies to bring extended functionality to existing BMI model implementations.  Please feel free to look over and leave thoughts/comments either here or in issues on the referenced repository.
